### PR TITLE
Use beat receiver subcomponent status for status translation

### PR DESCRIPTION
**Pull request by swiatekm**
## What does this PR do?

Uses subcomponent status metadata emitted by beats receivers to set unit and stream status for running components. This brings beat receiver status reporting up to par with beats processes.

As part of the implementation, I moved the otel status event serialization logic to its own package, so I can reuse it in status translation. Subcomponent statuses come in an untyped map and need to be converted to a status.Event struct before we can calculate the statuses.

This PR currently imports a patched status aggregation package from collector contrib, as the upstream PR adding the necessary functionality hasn't been merged yet. This uses our official fork and we'll have to maintain it through otel contrib upgrades until it lands and is released.

## Why is it important?

Beats receivers and beats processes should be at feature parity. In particular, individual units and streams should have their status set based on the underlying beats input or module, and not based on the aggregated status of the whole receiver. If a system/metrics module emits a diagnostic message about some targets being unreachable (because they're processes owned by another uses and agent isn't running as root, for example), then this should work as expected too with beats receivers.

## Checklist

- [x] I have read and understood the [pull request guidelines](https://github.com/elastic/elastic-agent/blob/main/CONTRIBUTING.md#pull-request-guidelines) of this project.
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- [x] I have added an integration test or an E2E test

## How to test this PR locally

Build agent locally and try the following config. You should see degraded statuses on the streams.

```
agent:
  logging:
    to_stderr: true
    level: error
  monitoring:
    enabled: false
inputs:
- data_stream:
    namespace: default
  id: unique-system-metrics-input
  streams:
  - data_stream:
      dataset: system.cpu
    metricsets:
    - cpu
    id: unique-system-metrics-input-cpu
  - data_stream:
      dataset: system.process
    metricsets:
    - process
    id: unique-system-metrics-input-process
  type: system/metrics
  use_output: default
- data_stream:
    namespace: default
  id: unique-system-metrics-input-2
  streams:
  - data_stream:
      dataset: system.cpu
    metricsets:
    - cpu
    id: unique-system-metrics-input-2-cpu
  - data_stream:
      dataset: system.process
    metricsets:
    - process
    id: unique-system-metrics-input-2-process
    degrade_on_partial: true
  type: system/metrics
  use_output: default
outputs:
  default:
    username: elastic
    password: elastic
    hosts:
    - 127.0.0.1:9200
    type: elasticsearch
    status_reporting:
      enabled: false
```

## Related issues

- Relates https://github.com/elastic/ingest-dev/issues/5742

### CODE REVIEW

File: pkg/component/component.go [hunk 1]
```diff
@@ -188,6 +188,17 @@ type Unit struct {
 	Err error `yaml:"error,omitempty"`
 }
 
+func (u *Unit) GetInputID(componentID string) string {
+	if u.Type == client.UnitTypeOutput {
+		return ""
+	}
+	inputID, found := strings.CutPrefix(fmt.Sprintf("%s-", componentID), u.ID)
```
Comment (cmacknz) [inline pkg/component/component.go]: I can't follow what this does.

Normally the component ID would be something like `system/metrics-default` (unless it's endpoint, then it's just `endpoint` after https://github.com/elastic/elastic-agent/pull/11740) and the unit ID would be a superset of that like `system/metrics-default-unique-system-metrics-input-2`.

So [CutPrefix(s, prefix string)](https://pkg.go.dev/strings#CutPrefix) would never find that the unit ID is a prefix of the component ID.

This code would evaluate as `strings.CutPrefix("system/metrics-default-", "system/metrics-default-unique-system-metrics-input-2`")` where the second argument is the prefix.

Are the arguments reversed?

Comment (cmacknz) [reply pkg/component/component.go]: Alternatively, this is just always returning `u.ID` because there is never a prefix match and that works where this is used.

File: internal/pkg/otel/translate/status_test.go [hunk 2]
```diff
@@ -757,6 +758,338 @@ func TestStateWithMessage(t *testing.T) {
...
+			assert.Equal(t, tt.expected.Message, result.Message)
+			assert.Equal(t, tt.expected.Payload, result.Payload)
+		})
+	}
+}
+
+func TestGetComponentUnitStateWithUnitInputStatus(t *testing.T) {
+	// Test the unit status from input attributes path
+	// GetInputID returns: strings.CutPrefix(componentID + "-", unitID)
+	// So with componentID = "test-comp" and unitID = "test", GetInputID returns "-comp-"
```
Comment (cmacknz) [inline internal/pkg/otel/translate/status_test.go]: > // So with componentID = "test-comp" and unitID = "test", GetInputID returns "-comp-"

What do we consider `-comp-` to be a correct return value from GetInputID?

File: pkg/component/component.go (hunk 1 above)
Comment (swiatekm) [reply pkg/component/component.go]: Yeah, the arguments are in the wrong order and it doesn't work at all. But the only case where it matters is where we have a filestream input without streams - the only beats input type allowed to do that. And even in that case, we just fall back to the component, so the only case where it would matter is a bit contrived. We definitely don't have an integration test for it.

I'll fix it and make sure it gives reasonable results for endpoint. Endpoint will never run as a beats receiver, so the status translation code is fine, but who knows what it will get used for in the future.

**Comment by swiatekm** [comment]
> Otherwise, LGTM. I doubt we are going to get the upstream PR merged before we want this to merge, so we can either go ahead with using our contrib fork or we can just keep using the private fork knowing it's temporary.

I'd much prefer moving it to our official fork. My fork isn't even under the elastic github org.

File: pkg/component/component.go (hunk 1 above)
Comment (cmacknz) [reply pkg/component/component.go]: Thanks, FWIW I also expected the arguments to be in the reverse order until I read the docs.

Comment (swiatekm) [reply pkg/component/component.go]: Fixed in https://github.com/elastic/elastic-agent/pull/11856/changes/8b651cbef50c0276a5d9b2876ce3219112845852. I ended up moving the method to the Component struct and scoping it aggressively so it only does anything for Beats input types.

File: internal/pkg/otel/translate/status_test.go (hunk 2 above)
Comment (swiatekm) [reply internal/pkg/otel/translate/status_test.go]: We don't. The perils of letting an LLM generate tests for "trivial" pure functions. :computer:

File: pkg/component/component.go (hunk 1 above)
Comment (swiatekm) [reply pkg/component/component.go]: I also did some manual tests for this edge case and everything worked correctly.

File: internal/pkg/otel/status/serializable.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+		err = errors.New(se.Error)
+	}
+
+	statusVal, ok := stringToStatusMap[se.StatusString]
+	if !ok {
+		statusVal = componentstatus.StatusNone
+	}
+
+	attributes := pcommon.NewMap()
+	// TODO: handle the error here
```
Comment (michel-laterman) [inline internal/pkg/otel/status/serializable.go]: Is this comment still relevant?

File: internal/pkg/otel/translate/status.go [hunk 4]
```diff
@@ -477,7 +477,7 @@ func getComponentUnitState(otelUnitStatus *status.AggregateStatus, unit componen
 
 	// get the unit status from input statuses
 	// this can happen for filestream, which is allowed to have a configuration without streams
-	unitInputId := unit.GetInputID(componentID)
+	unitInputId := comp.GetBeatInputIDForUnit(unit.ID)
```
Comment (cmacknz) [inline internal/pkg/otel/translate/status.go:480]: Why does this have to be Beat specific? We have tried to keep most of agent (except the self-monitoring stuff...) agnostic of what specific input it's running.

Comment (swiatekm) [reply internal/pkg/otel/translate/status.go:480]: It doesn't, but scoping it down to that case makes the logic simpler. If I want to properly implement this for all inputs, I need to not only handle all the cases (isolated vs non-isolated, service vs non-service, etc.), but also write unit tests verifying that this works for all those cases, so any future change won't break it. Which also forces me to refactor how the ids are determined for both components and units.

This PR is large enough, so I'd rather do this in a follow-up. But if you think it should go here, no problem.

Comment (cmacknz) [reply internal/pkg/otel/translate/status.go:480]: You won't have to handle the service runtime, it doesn't have streams. I think I can accept using this as a shortcut for other properties (e.g. input has or does not have a streams section) but I would want a comment explaining why so someone who is not either of us can understand how to refactor it later. I'm not sure I understand what is special about Beats here. They are the only inputs that use streams right now, but that isn't called out.

The comment says "input without streams" which is possible for all Beats but filestream is the one that shows up in our own examples.

Comment (swiatekm) [reply internal/pkg/otel/translate/status.go:480]: For clarity, this isn't about streams per se. What I need is the ability to recover the agent input ID from the unit and component. The reason I need this is that, if you create a filestream input without streams, then the stream id becomes the input id, effectively. But in principle, I'm just reversing the ID generation functions, so this could work for literally all input types. I'd just rather not do it completely in this PR.

> The comment says "input without streams" which is possible for all Beats but filestream is the one that shows up in our own examples.

Looks like this is not true:
https://github.com/elastic/beats/blob/03fff9b4b81f0ade8b7273bd62ff0df8bdf438db/x-pack/libbeat/management/generate.go#L83. We error if it's not filestream.

Comment (cmacknz) [reply internal/pkg/otel/translate/status.go:480]: > For clarity, this isn't about streams per se. What I need is the ability to recover the agent input ID from the unit and component. The reason I need this is that, if you create a filestream input without streams, then the stream id becomes the input id, effectively. But in principle, I'm just reversing the ID generation functions, so this could work for literally all input types. I'd just rather not do it completely in this PR.

Putting something like this in a comment would satisfy me here, specifically the part where there is no reason the function can't be generic except that the extra effort to generalize it. Right now it's hard to tell that there isn't some actual restriction this is working around.

> Looks like this is not true:
https://github.com/elastic/beats/blob/03fff9b4b81f0ade8b7273bd62ff0df8bdf438db/x-pack/libbeat/management/generate.go#L83. We error if it's not filestream.

Interesting, there's no reason other inputs can't support this, since streams is an agent concept not a beats concept.

Comment (swiatekm) [reply internal/pkg/otel/translate/status.go:480]: Added more comments here: https://github.com/elastic/elastic-agent/pull/11856/changes/9a2abee5668eac6911e38efecb4c1b183fbf35b8.

**Review by cmacknz — APPROVED** [review]

File: internal/pkg/otel/status/serializable.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+	statusVal, ok := stringToStatusMap[se.StatusString]
+	if !ok {
+		statusVal = componentstatus.StatusNone
+	}
+
+	attributes := pcommon.NewMap()
+	parseErr := attributes.FromRaw(se.Attributes)
+	if parseErr != nil {
+		panic(fmt.Errorf("error parsing event attributes %v: %w", se.Attributes, parseErr))
```
Comment (cmacknz) [inline internal/pkg/otel/status/serializable.go]: This needs to be an error return not a panic. An invalid attribute cannot cause agent to exit.

Comment (swiatekm) [reply internal/pkg/otel/status/serializable.go]: Fixed: https://github.com/elastic/elastic-agent/pull/11856/changes/fbd0d3f15c652d586a64413f228ec6881c5c276a. I swore I fixed this before, but apparently not. This can't realistically happen unless there's a bug in the healthcheck extension, but there's no reason to not handle the error properly.

**Comment by cmacknz** [comment]
Notice needs to be updated, otherwise LGTM (again - this one is cursed apparently)
